### PR TITLE
Fixed #27704 -- Fixed TypedMultipleChoiceField as base field class for contrib.postgres.ArrayField with choices.

### PR DESCRIPTION
**Pull request by Anv3sh**
As per suggestions in @ngnpope's [comment](https://github.com/django/django/pull/7850#issuecomment-738300243) on #7850 made `choices_form_class = TypedMultipleChoiceField` also removed support for `ArrayField.choices` and instead `ArrayField.base_field.choices` shall be used. Removed `ArrayField.clean()` method intoduced in https://github.com/django/django/commit/9dd244394236388c3479ab202a0ec31055f7ec09 because `SimpleArrayField` will no more be used while `ArrayField` has choices and also shifted the test `test_model_field_choices` introduced in https://github.com/django/django/commit/9dd244394236388c3479ab202a0ec31055f7ec09 to `tests.postgres_tests.test_array.TestChoiceFormField`.

### CODE REVIEW

**Comment by felixxm** [comment]
@ngnpope Do you want to take a look?

**Review by ngnpope — COMMENTED** [review]
Thanks @felixxm. Haven't got all that much time at the moment, but have given this a quick look.

The main question is whether we're going to deprecate and remove support for `choices` on `ArrayField`. It probably makes sense. In which case this also ought to grow a system check, initially as a warning, to be promoted to error on removal.

File: django/contrib/postgres/fields/array.py [hunk 1]
```diff
@@ -202,14 +203,23 @@ def run_validators(self, value):
...
                 "form_class": SimpleArrayField,
                 "base_field": self.base_field.formfield(),
                 "max_length": self.size,
-                **kwargs,
             }
-        )
+            defaults.update(kwargs)
+        else:
+            raise NotImplementedError("Choices should be defined in base field.")
+        return super().formfield(**defaults)
```
[the remark this answers is not in the document]
Comment (Anv3sh) [reply django/contrib/postgres/fields/array.py:222]: Didn't get idea of having the obj variable? 🤔

[the remark this answers is not in the document]
Comment (ngnpope) [reply django/contrib/postgres/fields/array.py:222]: Sorry - typo. Fixed.

[the remark this answers is not in the document]
Comment (Anv3sh) [reply django/contrib/postgres/fields/array.py:222]: So got through this snippet and build a test case for it as follows which successfully passed:
```py
    model_field = ArrayField(
            models.CharField(max_length=27), choices=(("a1", "A1"), ("b1", "B1"))
        )
        with warnings.catch_warnings(record=True) as w:
            warnings.simplefilter("always")
            model_field.formfield()
            assert len(w) == 1
            assert issubclass(w[-1].category, RemovedInDjango51Warning)
            assert "Choices should be defined in base field." in str(w[-1].message)
```
but adding objects through the admin page ignored the warning with the `form_class` being `TypedChoiceField` and saving the object gave the validation error stated in this [comment](https://github.com/django/django/pull/7850#discussion_r101889434).

**Comment by Anv3sh** [comment]
Committed the changes according to @ngnpope's review and added test for it although we still need to look for a fix for some shortcomings that I picked out in this [comment](https://github.com/django/django/pull/15805#discussion_r921342795).

**Comment by Anv3sh** [comment]
@felixxm if someone could review 🙇🏻

**Comment by Anv3sh** [comment]
@ngnpope can you have a look??

File: django/contrib/postgres/fields/array.py [hunk 2]
```diff
@@ -202,14 +205,23 @@ def run_validators(self, value):
...
+            defaults = {
                 "form_class": SimpleArrayField,
                 "base_field": self.base_field.formfield(),
                 "max_length": self.size,
-                **kwargs,
             }
-        )
+            if self.choices:
+                warnings.warn(
+                    "Choices should be defined in base field.", RemovedInDjango51Warning
```
Comment (jacobtylerwalls) [inline django/contrib/postgres/fields/array.py:222]:
```suggestion
                    "Choices should be defined in base field.", RemovedInDjango51Warning, stacklevel=2
```

File: tests/postgres_tests/test_array.py [hunk 3]
```diff
@@ -1113,6 +1110,43 @@ def test_has_changed_empty(self):
...
+    def test_model_field_formfield_choices(self):
+        model_field = ArrayField(
+            models.CharField(max_length=27), choices=(("a1", "A1"), ("b1", "B1"))
+        )
+        with warnings.catch_warnings(record=True) as w:
+            warnings.simplefilter("always")
+            model_field.formfield()
+            assert len(w) == 1
+            assert issubclass(w[-1].category, RemovedInDjango51Warning)
+            assert "Choices should be defined in base field." in str(w[-1].message)
```
Comment (jacobtylerwalls) [inline tests/postgres_tests/test_array.py:1147]:
```suggestion
        msg = "Choices should be defined in base field."
        with self.assertWarnsMessage(RemovedInDjango51Warning, msg):
            model_field.formfield()
```

**Comment by Anv3sh** [comment]
@jacobtylerwalls committed the suggested changes although the patch does not throws a warning if choices are not in `base field` as far as i tested even though we have that in the patch and also the test works fine.

**Comment by jacobtylerwalls** [comment]
Ah, along the lines of this https://github.com/django/django/pull/15805#discussion_r921342795?

You could consider changing the warning to a system check as Nick [suggested](https://github.com/django/django/pull/15805#pullrequestreview-1035447454).

**Comment by Anv3sh** [comment]
> Ah, along the lines of this [#15805 (comment)](https://github.com/django/django/pull/15805#discussion_r921342795)?
> 
> You could consider changing the warning to a system check as Nick [suggested](https://github.com/django/django/pull/15805#pullrequestreview-1035447454).
Got it 👍🏻  will look into it

**Comment by Anv3sh** [comment]
> Ah, along the lines of this [#15805 (comment)](https://github.com/django/django/pull/15805#discussion_r921342795)?
> 
> You could consider changing the warning to a system check as Nick [suggested](https://github.com/django/django/pull/15805#pullrequestreview-1035447454).

So we would have to raise an error on running the `makemigrations` command right?
Also according to our current change that wouldn't be possible cause we would only be able to check it once the `ArrayField.formfield()` is called.

**Comment by ngnpope** [comment]
Have a look at the [documentation](https://docs.djangoproject.com/en/4.1/topics/checks/#field-model-manager-and-database-checks) for field checks. This is what I was referring to.

**Comment by felixxm** [comment]
Closing due to inactivity.

**Pull request closed without merging**